x/gamm/twap/logic_test.go

```diff
@@ -0,0 +1,206 @@
+package twap_test
+
+import (
+	"testing"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/stretchr/testify/require"
+
+	"github.com/osmosis-labs/osmosis/v10/osmoutils"
+	"github.com/osmosis-labs/osmosis/v10/x/gamm/twap"
+	"github.com/osmosis-labs/osmosis/v10/x/gamm/twap/types"
+)
+
+var zeroDec = sdk.ZeroDec()
+var oneDec = sdk.OneDec()
+var twoDec = oneDec.Add(oneDec)
+var OneSec = sdk.NewDec(1e9)
+
+func newRecord(t time.Time, sp0, accum0, accum1 sdk.Dec) types.TwapRecord {
+	return types.TwapRecord{
+		Asset0Denom:     defaultUniV2Coins[0].Denom,
+		Asset1Denom:     defaultUniV2Coins[1].Denom,
+		Time:            t,
+		P0LastSpotPrice: sp0,
+		P1LastSpotPrice: sdk.OneDec().Quo(sp0),
+		// make new copies
+		P0ArithmeticTwapAccumulator: accum0.Add(sdk.ZeroDec()),
+		P1ArithmeticTwapAccumulator: accum1.Add(sdk.ZeroDec()),
+	}
+}
+
+// make an expected record for math tests, we adjust other values in the test runner.
+func newExpRecord(accum0, accum1 sdk.Dec) types.TwapRecord {
+	return types.TwapRecord{
+		Asset0Denom: defaultUniV2Coins[0].Denom,
+		Asset1Denom: defaultUniV2Coins[1].Denom,
+		// make new copies
+		P0ArithmeticTwapAccumulator: accum0.Add(sdk.ZeroDec()),
+		P1ArithmeticTwapAccumulator: accum1.Add(sdk.ZeroDec()),
+	}
+}
+
+func TestInterpolateRecord(t *testing.T) {
+	tests := map[string]struct {
+		record          types.TwapRecord
+		interpolateTime time.Time
+		expRecord       types.TwapRecord
+	}{
+		"0accum": {
+			record:          newRecord(time.Unix(1, 0), sdk.NewDec(10), zeroDec, zeroDec),
+			interpolateTime: time.Unix(2, 0),
+			expRecord:       newExpRecord(OneSec.MulInt64(10), OneSec.QuoInt64(10)),
+		},
+		"small starting accumulators": {
+			record:          newRecord(time.Unix(1, 0), sdk.NewDec(10), oneDec, twoDec),
+			interpolateTime: time.Unix(2, 0),
+			expRecord:       newExpRecord(oneDec.Add(OneSec.MulInt64(10)), twoDec.Add(OneSec.QuoInt64(10))),
+		},
+		"larger time interval": {
+			record:          newRecord(time.Unix(11, 0), sdk.NewDec(10), oneDec, twoDec),
+			interpolateTime: time.Unix(55, 0),
+			expRecord:       newExpRecord(oneDec.Add(OneSec.MulInt64(44*10)), twoDec.Add(OneSec.MulInt64(44).QuoInt64(10))),
+		},
+		"same time": {
```

Comment:
What happens if `newTime` is before record time by mistake so the `timeDelta` is negative? How is this handled by the core logic? Is there a test covering this?

Reply:
Hrmm, there isn't a test handling it because its incorrect behavior. But you should still get the same result. This is because TWAP is:
```
(accumEnd - accumStart) / (timeEnd - timeStart)
```

So if you reverse the arguments, you compute 
```
(accumStart - accumEnd) / (timeStart - timeEnd)
= -(accumEnd - accumStart) / - (timeEnd - timeStart)
= (accumEnd - accumStart) / (timeEnd - timeStart)
```
Thus the function is argument order independent.

This is for the same reason that you can don't care about the order of points when trying to compute the slope of a line.

Reply:
That makes sense, thanks for explaining. I think it still makes to add a test even if the behavior is incorrect. We should aim to have test cases relative to all possible inputs even though some might be invalid.